core/src/main/java/org/elasticsearch/index/IndexSortConfig.java

```diff
@@ -0,0 +1,218 @@
+                throw new IllegalArgumentException("index.sort.field:" + Arrays.toString(fields) +
+                    " index.sort.order:" + Arrays.toString(orders) + ", size mismatch");
+            }
+        } else {
+            orders = new SortOrder[fields.length];
+        }
+
+        if (INDEX_SORT_MODE_SETTING.exists(settings)) {
+            modes = INDEX_SORT_MODE_SETTING.get(settings)
+                .toArray(new MultiValueMode[fields.length]);
+            if (modes.length != fields.length) {
+                throw new IllegalArgumentException("index.sort.field:" + Arrays.toString(fields) +
+                    " index.sort.mode:" + Arrays.toString(modes) + ", size mismatch");
+            }
+        } else {
+            modes = new MultiValueMode[fields.length];
+        }
+
+        if (INDEX_SORT_MISSING_SETTING.exists(settings)) {
+            missingValues = INDEX_SORT_MISSING_SETTING.get(settings)
+                .toArray(new String[fields.length]);
+            if (missingValues.length != fields.length) {
+                throw new IllegalArgumentException("index.sort.field:" + Arrays.toString(fields) +
+                    " index.sort.missing:" + Arrays.toString(missingValues) + ", size mismatch");
+            }
+        } else {
+            missingValues = new String[fields.length];
+        }
+    }
+
+
+    /**
+     * Returns true if the index should be sorted
+     */
+    public boolean hasIndexSort() {
+        return fields.length > 0;
+    }
+
+    /**
+     * Builds the {@link Sort} order from the settings for this index
+     * or returns null if this index has no sort.
+     */
+    public Sort buildIndexSort(Function<String, MappedFieldType> fieldTypeLookup,
+                               Function<MappedFieldType, IndexFieldData<?>> fieldDataLookup) {
+        if (hasIndexSort() == false) {
+            return null;
+        }
+
+        final SortField[] sortFields = new SortField[fields.length];
+        for (int i = 0; i < fields.length; i++) {
+            final MappedFieldType ft = fieldTypeLookup.apply(fields[i]);
+            if (ft == null) {
+                throw new IllegalArgumentException("unknown index sort field:[" + fields[i] + "]");
+            }
+            boolean reverse = orders[i] == null ? false : (orders[i] == SortOrder.DESC);
+            MultiValueMode mode =
+                modes[i] == null ?
+                    reverse ? MultiValueMode.MAX : MultiValueMode.MIN :
+                    modes[i];
+            IndexFieldData<?> fieldData;
+            try {
+                fieldData = fieldDataLookup.apply(ft);
+            } catch (Exception e) {
+                throw new IllegalArgumentException("can't load docvalues for index sort field:[" + fields[i] + "]");
+            }
+            if (fieldData == null) {
+                throw new IllegalArgumentException("fielddata not found for index sort field:[" + fields[i] + "]");
```

Comment:
Should this be `docvalues not found for index sort field [`? Like, without the `:` and s/fielddata/docvalues/. I know we abstract docvalues behind something we call fielddata in the code base but in the docs we mostly talk about fielddata as that uninverted thing that eats up all your heap.